### ISSUE TITLE
[Client bug]: BadRequest error: Parsing OData Select and Expand failed: Syntax error: character '+' is not valid at position 

### ISSUE THREAD

**Issue by twieczor** [microsoftgraph/msgraph-sdk-dotnet#1081]
**Describe the bug**
After upgrading to version 4.0.0 graph requests with the following expand no longer work:

singleValueExtendedProperties($filter=Id eq 'String 0x001A')

The graph client builds and sends to the endpoint the following request:

GET /v1.0/me/mailFolders('Inbox')/messages?$filter=isRead+ne+true&$select=BodyPreview,ChangeKey,Flag,From,HasAttachments,Id,Importance,InferenceClassification,IsDraft,IsRead,IsReadReceiptRequested,LastModifiedDateTime,ReceivedDateTime,Sender,SentDateTime,Subject,WebLink&$expand=singleValueExtendedProperties($filter=Id+eq+'String+0x001A')&$orderby=receivedDateTime&$top=10&$skip=10

And receives error:

{"error":{"code":"BadRequest","message":"Parsing OData Select and Expand failed: Syntax error: character '+' is not valid at position 2 in 'Id+eq+'String+0x001A''.","innerError":{"date":"2021-07-29T13:54:45","request-id":"7f617e4d-e5cf-47fc-ae7c-8ef602b4818d","client-request-id":"7f617e4d-e5cf-47fc-ae7c-8ef602b4818d"}}}

It looks like the + character in place of spaces is rejected. The '+' character is inserted by the library wherever there are spaces in the filter.

**Expected behavior**
Graph query returns response

**Screenshots**

**Client version**
Microsoft Graph .NET SDK v4

**Desktop (please complete the following information):**

**Additional context**
 
[AB#10439](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/10439)

**Comment by twieczor** [comment]
After digging a little deeper it looks like the repro is more complex. There are 2 request. The first request works, and then when we page, the page request fails.

First request code:

IMailFolderMessagesCollectionRequestBuilder miv = null;

miv = clientGraph.Me.MailFolders.Inbox.Messages;
var page = await miv.Request()
.Filter(filter)                .Select("BodyPreview,ChangeKey,Flag,From,HasAttachments,Id,Importance,InferenceClassification,IsDraft,IsRead,IsReadReceiptRequested,LastModifiedDateTime,ReceivedDateTime,Sender,SentDateTime,Subject,WebLink")
                    .Expand("SingleValueExtendedProperties($filter=Id eq 'String 0x001A')")
                    .OrderBy("receivedDateTime")
                    .GetAsync(cancelToken);

The first request produces the following request:

https://graph.microsoft.com/v1.0/me/mailFolders/Inbox/messages?$filter=isRead%20ne%20true&$select=BodyPreview%2CChangeKey%2CFlag%2CFrom%2CHasAttachments%2CId%2CImportance%2CInferenceClassification%2CIsDraft%2CIsRead%2CIsReadReceiptRequested%2CLastModifiedDateTime%2CReceivedDateTime%2CSender%2CSentDateTime%2CSubject%2CWebLink&$expand=SingleValueExtendedProperties%28%24filter%3DId%20eq%20%27String%200x001A%27%29&$orderby=receivedDateTime

The first request returns valid response:

{"@odata.context":"https://graph.microsoft.com/v1.0/$metadata#users('afbb3c2d-a6e2-4d09-8633-933b71187b72')/mailFolders('Inbox')/messages(bodyPreview,changeKey,flag,from,hasAttachments,id,importance,inferenceClassification,isDraft,isRead,isReadReceiptRequested,lastModifiedDateTime,receivedDateTime,sender,sentDateTime,subject,webLink,singleValueExtendedProperties())","value":[{"@odata.etag":"W/\"FwAAABYAAABp16Ol/VTPS7l56vY+TljuAAANpW5U\"","id":"AAMkADllNzVkY2JlLTkwYTQtNDk1NS1hMzU1LWI0Y2UxZjdjYWM1ZABGAAAAAAAedlJuQnW9QpW_uw6LYKcOBwBp16Ol-VTPS7l56vY_TljuAAAAAAEMAABp16Ol-VTPS7l56vY_TljuAAANqWjQAAA=","lastModifiedDateTime":"2019-12-04T19:25:11Z","changeKey":"FwAAABYAAABp16Ol/VTPS7l56vY+TljuAAANpW5U","receivedDateTime":"2019-12-03T17:10:01Z","sentDateTime":"2019-12-03T17:09:57Z","hasAttachments":false,"subject":"Undeliverable: [309-2579516C-0205] test email","bodyPreview":"Delivery .........................." ]}],
"@odata.nextLink":"https://graph.microsoft.com/v1.0/me/mailFolders('Inbox')/messages?%24filter=isRead+ne+true&%24select=BodyPreview%2cChangeKey%2cFlag%2cFrom%2cHasAttachments%2cId%2cImportance%2cInferenceClassification%2cIsDraft%2cIsRead%2cIsReadReceiptRequested%2cLastModifiedDateTime%2cReceivedDateTime%2cSender%2cSentDateTime%2cSubject%2cWebLink&%24expand=SingleValueExtendedProperties(%24filter%3dId+eq+%27String+0x001A%27)&%24orderby=receivedDateTime&%24top=10&%24skip=10"}

Then we try to page:

page = await page.NextPageRequest.GetAsync(cancelToken);

The page request produces the following request:

https://graph.microsoft.com/v1.0/me/mailFolders('Inbox')/messages?%24filter=isRead%2Bne%2Btrue&%24select=BodyPreview%2CChangeKey%2CFlag%2CFrom%2CHasAttachments%2CId%2CImportance%2CInferenceClassification%2CIsDraft%2CIsRead%2CIsReadReceiptRequested%2CLastModifiedDateTime%2CReceivedDateTime%2CSender%2CSentDateTime%2CSubject%2CWebLink&%24expand=SingleValueExtendedProperties%28%24filter%3DId%2Beq%2B%27String%2B0x001A%27%29&%24orderby=receivedDateTime&%24top=10&%24skip=10

Which returns error:

{"error":{"code":"BadRequest","message":"Parsing OData Select and Expand failed: Syntax error: character '+' is not valid at position 2 in 'Id+eq+'String+0x001A''.","innerError":{"date":"2021-07-29T15:55:34","request-id":"58a2ab17-0e35-4625-b4bd-76b6bf7d86b5","client-request-id":"58a2ab17-0e35-4625-b4bd-76b6bf7d86b5"}}}

**Comment by andrueastman** [comment]
@twieczor Thanks for bringing this issue to our attention.

It seems that in the response from the API the "@odata.nextLink" property has the space characters replaced with `+` characters which causes the issue as the `+` will be encoded by the library. 

For now, you should be able to work around this by using the code below as we investigate a way to mitigate this.

```cs
var originalRequest = miv.Request()
    .Filter(filter).Select("BodyPreview,ChangeKey,Flag,From,HasAttachments,Id,Importance,InferenceClassification,IsDraft,IsRead,IsReadReceiptRequested,LastModifiedDateTime,ReceivedDateTime,Sender,SentDateTime,Subject,WebLink")
    .Expand("SingleValueExtendedProperties($filter=Id eq 'String 0x001A')")
    .OrderBy("receivedDateTime");

var page = await originalRequest.GetAsync(cancelToken);

// Replace filter clause from server with our original filter clause
var newFilter = page.NextPageRequest?.QueryOptions.FirstOrDefault(x => x.Name.EndsWith("filter"));
if (newFilter != null)
{
    var index = page.NextPageRequest.QueryOptions.IndexOf(newFilter);
    page.NextPageRequest.QueryOptions[index] = originalRequest.QueryOptions.FirstOrDefault(x => x.Name.EndsWith("filter"));
}

var nextPage = await page.NextPageRequest.GetAsync(cancelToken);
```

**Comment by Licantrop0** [comment]
this is also occurring on a simple select clause:
![image](https://user-images.githubusercontent.com/2388430/127764174-654bf6e4-6f16-4178-931d-a326587cf52a.png)

(the code was perfectly working in the previous version)

**Comment by andrueastman** [comment]
@Licantrop0 Is there a specific reason that your select clause has the plus sign?

My understanding is that it should be something like this with the property names being simply comma separated.

```cs
var users = await graphServiceClient.Users.Request().Select("id,displayName").GetAsync();
```

**Comment by ryan-carbon** [comment]
I'm having a similar problem with .OrderBy("lastDeliveredDateTime desc") 



            Graph.IGraphServiceGroupsCollectionPage groupsCollectionPage = await graphClient.Groups
                .Request()
                .Select("id")
                .Filter($"displayName eq '{settings.EmailGroup}'")
                .GetAsync();

            Graph.Group group = groupsCollectionPage
                .CurrentPage
                .Single();

            Graph.IGroupThreadsCollectionRequestBuilder builder = graphClient.Groups[group.Id].Threads;
            Graph.IGroupThreadsCollectionPage threads = await builder.Request()
                .Select("id,topic,lastDeliveredDateTime,hasAttachments,uniqueSenders")
                .Expand("Posts($expand=Attachments($select=id)),Posts($select=id,from,createdDateTime)")
                .OrderBy("lastDeliveredDateTime desc")
                .GetAsync();

            threads = await threads.NextPageRequest.GetAsync();

In 3.35.0 the last line produces 

https://graph.microsoft.com/v1.0/groups/xx/threads?%24select=id%2ctopic%2clastDeliveredDateTime%2chasAttachments%2cuniqueSenders&%24expand=Posts(%24expand%3dAttachments(%24select%3did))%2cPosts(%24select%3did%2cfrom%2ccreatedDateTime)&%24orderby=lastDeliveredDateTime+desc&%24skip=10

vs. 4.0.0

https://graph.microsoft.com/v1.0/groups/f8f48434-8ee4-40f6-a758-57321549cb04/threads?%24select=id%2Ctopic%2ClastDeliveredDateTime%2ChasAttachments%2CuniqueSenders&%24expand=Posts%28%24expand%3DAttachments%28%24select%3Did%29%29%2CPosts%28%24select%3Did%2Cfrom%2CcreatedDateTime%29&%24orderby=lastDeliveredDateTime%2Bdesc&%24skip=10

**Comment by ghelyar** [comment]
This should really be given a high priority. It makes 4.x basically unusable, and it's been broken for almost 2 weeks now.

Here's a simple repro:
```csharp
var client = new GraphServiceClient(authProvider);

// https://graph.microsoft.com/v1.0/users?$filter=accountEnabled eq true&$top=1
var request1 = client.Users.Request().Filter("accountEnabled eq true").Top(1);
var result1 = await request1.GetAsync();

// https://graph.microsoft.com/v1.0/users?$filter=accountEnabled%2Beq%2Btrue&$top=1&$skiptoken=...
var request2 = result1.NextPageRequest;
var result2 = await request2.GetAsync(); // throws ServiceException Code: BadRequest Message: Invalid filter clause
```

Note that `request2.QueryOptions["$filter"] == "accountEnabled+eq+true"` in both the 3.x and 4.x versions of the library.

[Here is the commit that broke it](https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/commit/1ae669a9c668c7310360bbf98c0110511a05c879)

```
var unescapedData = Uri.UnescapeDataString("accountEnabled+eq+true"); // accountEnabled+eq+true
var escapedOptionValue = Uri.EscapeDataString(unescapedData); // accountEnabled%2Beq%2Btrue
```

I don't know why you unescape and then immediately escape again, that doesn't seem right, but note that `System.Uri.UnescapeDataString` converts `%20` to ` ` but _does not_ convert `+` back to ` `. Also note that `System.Net.WebUtility.UrlDecode` _does_ convert `+` back to ` ` so you could fix this just by changing this from `System.Uri.UnescapeDataString` to `System.Net.WebUtility.UrlDecode`. A well placed `string.Replace('+', ' ')` would also probably work, but you would have to do it before the `Uri.UnescapeDataString`. Note that when URL encoded, both `+` and `%20` mean space.

**Comment by andrueastman** [comment]
Hey @ghelyar. 

Thanks for the tip!

Unfortunately, it will be tricky to simply convert/replace `+` with space characters as that would break valid use cases where queries have `+` characters in them. As seen in https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/700 and https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/914

As this specifically affects the initialization of the `NextPageRequest` I believe it would be better to fix it there (especially since the url returned differs from the one sent initially, and the issue is caused there). 
We are currently, working towards fixing this and this should be resolved in the next update of the SDK in the coming week.

**Comment by ghelyar** [comment]
OK, but what you've actually done is broken other valid use cases where % means % and where + means space, and that's broken paging, which is a core part of the library, for an edge case that could have previously been achieved by URL encoding the filter in the caller.

The "proper" fix is to make it clear which strings should be URL encoded, and which strings should not, because `+` means a different thing depending on whether the input is already URL encoded, to if it isn't.

* If the query options should be URL encoded by the caller, then no decode/encode should happen in `RequestBase.BuildQueryString`, the breaking commit should be reverted, and the caller should be told to URL encode it themselves in the docs e.g. `request.Filter(WebUtility.UrlEncode("a + b"))`
* If the query options should _not_ be URL encoded by the caller, then `RequestBase.BuildQueryString` should _only_ `encode`, and _not_ decode first, the caller should be told _not_ to URL encode in the docs, and `NextPageRequest` should also follow that by decoding the URL itself.

Having it so that the input string *might* be URL encoded or *might not* be URL encoded can never work for all cases, because the same characters have different meanings depending on whether it's already encoded or not. You will always fix someone and break someone else if you do it this way.

**Comment by andrueastman** [comment]
Thanks for your patience, everyone.
This is now resolved via the [4.2.0 release](https://www.nuget.org/packages/Microsoft.Graph/4.2.0) and therefore closing this for now.
